pkg/controller/utils/labelselector.go

```diff
@@ -0,0 +1,50 @@
+package utils
+
+import (
+	"github.com/go-logr/logr"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+)
+
+// LabelSelector2LabelSelector converts a "k8s.io/apimachinery/pkg/apis/meta/v1".LabelSelector as found in manifests spec section into a "k8s.io/apimachinery/pkg/labels".Selector to be used to filter list operations.
+func LabelSelector2LabelSelector(logger logr.Logger, inSelector *metav1.LabelSelector) (outSelector labels.Selector, err error) {
+	outSelector = labels.NewSelector()
+	if inSelector != nil {
+		for key, value := range inSelector.MatchLabels {
+			req, err2 := labels.NewRequirement(key, selection.In, []string{value})
+			if err2 != nil {
+				logger.Error(err, "NewRequirement")
+				err = err2
+				continue
+			}
+			outSelector = outSelector.Add(*req)
+		}
+
+		for _, expr := range inSelector.MatchExpressions {
+			var op selection.Operator
+			switch expr.Operator {
+			case metav1.LabelSelectorOpIn:
+				op = selection.In
+			case metav1.LabelSelectorOpNotIn:
+				op = selection.NotIn
+			case metav1.LabelSelectorOpExists:
+				op = selection.Exists
+			case metav1.LabelSelectorOpDoesNotExist:
+				op = selection.DoesNotExist
+			default:
+				logger.Info("Invalid Operator:", expr.Operator)
+				continue
+			}
+			req, err2 := labels.NewRequirement(expr.Key, op, expr.Values)
+			if err2 != nil {
+				logger.Error(err, "NewRequirement")
+				err = err2
+				continue
+			}
+			outSelector = outSelector.Add(*req)
+		}
+	}
+	return outSelector, err
```

Comment:
it looks like even though we continue through the loops when there is an error here, we `return` on error when `LabelSelector2LabelSelector` is called (e.g. https://github.com/DataDog/extendeddaemonset/pull/25/files#diff-6d4acc3f7e4675ba93d62c645e490c0eR370) or we don't use the returned selector (https://github.com/DataDog/extendeddaemonset/pull/25/files?file-filters%5B%5D=.go#diff-6f8ca7266c71f02fdcdcc35fb5f12acaR294). is it worth continue the loop in these cases?

Reply:
You’re right.
I suppose it’s clearer to generate an error rather than silently ignoring only the part of a selector which is not valid.
I’ve changed to use early `return`s.